crates/oxc_linter/src/rules/unicorn/no_useless_spread/mod.rs

```diff
@@ -326,153 +338,141 @@ fn check_useless_iterable_to_array<'a>(
             {
-                return;
+                return false;
             }
 
             ctx.diagnostic_with_fix(
                 iterable_to_array(
                     span,
                     &get_method_name(call_expr).unwrap_or_else(|| "unknown".into()),
                 ),
-                |fixer| fix_by_removing_spread(fixer, array_expr, spread_elem),
+                |fixer| fix_by_removing_array_spread(fixer, array_expr, spread_elem),
             );
+            true
         }
-        _ => {}
+        _ => false,
     }
 }
 
-fn check_useless_array_clone<'a>(array_expr: &ArrayExpression<'a>, ctx: &LintContext<'a>) {
-    if !is_single_array_spread(array_expr) {
-        return;
-    }
-
-    let ArrayExpressionElement::SpreadElement(spread_elem) = &array_expr.elements[0] else {
-        return;
-    };
+/// Matches `new {Set,WeakSet,Map,WeakMap}(iterable)`
+pub fn is_new_map_or_set_with_iterable(new_expr: &NewExpression) -> bool {
+    is_new_expression(new_expr, &["Map", "WeakMap", "Set", "WeakSet"], Some(1), Some(1))
+}
 
+fn check_useless_clone<'a>(
+    array_or_obj_span: Span,
+    is_array: bool,
+    spread_elem: &SpreadElement<'a>,
+    ctx: &LintContext<'a>,
+) -> bool {
     let span = Span::new(spread_elem.span.start, spread_elem.span.start + 3);
+    let target = spread_elem.argument.get_inner_expression();
 
-    if let Expression::CallExpression(call_expr) = &spread_elem.argument {
-        if !(is_method_call(
-            call_expr,
-            None,
-            Some(&[
-                "concat",
-                "copyWithin",
-                "filter",
-                "flat",
-                "flatMap",
-                "map",
-                "slice",
-                "splice",
-                "toReversed",
-                "toSorted",
-                "toSpliced",
-                "with",
-            ]),
-            None,
-            None,
-        ) || is_method_call(call_expr, None, Some(&["split"]), None, None)
-            || is_method_call(call_expr, Some(&["Object"]), Some(&["keys", "values"]), None, None)
-            || is_method_call(call_expr, Some(&["Array"]), Some(&["from", "of"]), None, None))
-        {
-            return;
-        }
-
-        let method = call_expr.callee.get_member_expr().map_or_else(
-            || "unknown".into(),
-            |method| {
-                let object_name = if let Expression::Identifier(ident) = &method.object() {
-                    ident.name.as_str()
-                } else {
-                    "unknown"
-                };
+    // already diagnosed by first check
+    if matches!(target, Expression::ArrayExpression(_) | Expression::ObjectExpression(_)) {
+        return false;
+    }
 
-                format!("{}.{}", object_name, method.static_property_name().unwrap())
-            },
-        );
+    let hint = target.const_eval();
+    let hint_matches_expr = if is_array { hint.is_array() } else { hint.is_object() };
```

Comment:
What's the means of `hint_matches_expr` here? 🤔

Reply:
true if
1. we're checking for clones on arrays and `ValueHint` suggests the value is an array
2. vice versa for objects